pkg/backends/azurekeyvault.go

```diff
@@ -0,0 +1,78 @@
+package backends
+
+import (
+	"context"
+	"fmt"
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/keyvault/keyvault"
+	"path"
+	"strings"
+	"time"
+)
+
+// AzureKeyVault is a struct for working with an Azure Key Vault backend
+type AzureKeyVault struct {
+	Client keyvault.BaseClient
+}
+
+// NewAzureKeyVaultBackend initializes a new Azure Key Vault backend
+func NewAzureKeyVaultBackend(client keyvault.BaseClient) *AzureKeyVault {
+	return &AzureKeyVault{
+		Client: client,
+	}
+}
+
+// Login does nothing as a "login" is handled on the instantiation of the Azure SDK
+func (a *AzureKeyVault) Login() error {
+	return nil
+}
+
+// GetSecrets gets secrets from Azure Key Vault and returns the formatted data
+// For Azure Key Vault, `kvpath` is the unique name of your vault
+func (a *AzureKeyVault) GetSecrets(kvpath string, version string, _ map[string]string) (map[string]interface{}, error) {
+	kvpath = fmt.Sprintf("https://%s.vault.azure.net", kvpath)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	data := make(map[string]interface{})
+
+	secretList, err := a.Client.GetSecrets(ctx, kvpath, nil)
+	if err != nil {
+		return nil, err
+	}
+	// Gather all secrets in Key Vault
+	for ; secretList.NotDone(); secretList.NextWithContext(ctx) {
```

Comment:
Lets say you have a kubernetes object like:

```yaml
kind: Secret
apiVersion: v1
metadata:
  name: example-secret
type: Opaque
data:
  password: <path:josannkv#foo>
  password-2: <path:josannkv#josansecret>
  password-3: <path:josannkv#foo#8f8da2e06c8240808ee439ff093803b5>
  password-4: <path:josannkv#josansecret#33740fc26214497f8904d93f20f7db6d>
  password-5: <path:josannkv#foo#4509bc1bbc2341ca8e195b23f829af01>
```

1. So lets say the secret list consists of `josansecret` then followed by `foo`. 

2. Version `8f8da2e06c8240808ee439ff093803b5` gets passed in from password-3 above.

3. The for loop would fail since `josansecret` does not contain `8f8da2e06c8240808ee439ff093803b5`, but the next interation ( `foo` ) would have it but the for loop never gets to that point right breaking the secret lookup.

Or am I thinking wrong about this?

Reply:
Yeah, i guess you are right, I was thinking about it with the mindset of each inline doing a `GetSecret` with the version and not thinking about each secret in the list of secrets trying to retrieved with that version. 

Ideally, we would not need to loop the secrets but instead do a simple lookup to get one secret by name and version for the inlines paths. But i think the the project is written it doesn't allow for this right now.

Reply:
Yeah, its always possible to improve that in the future and if so then also fix this secret <-> version mapping.